consensus/beacon/consensus.go

```diff
@@ -256,8 +256,27 @@ func (beacon *Beacon) verifyHeader(chain consensus.ChainHeaderReader, header, pa
 	if diff := new(big.Int).Sub(header.Number, parent.Number); diff.Cmp(common.Big1) != 0 {
 		return consensus.ErrInvalidNumber
 	}
-	// Verify the header's EIP-1559 attributes.
-	return misc.VerifyEip1559Header(chain.Config(), parent, header)
+	if chain.Config().IsLondon(header.Number) {
+		// Verify the header's EIP-1559 attributes.
+		if err := misc.VerifyEip1559Header(chain.Config(), parent, header); err != nil {
+			return err
+		}
+	}
+	shanghai := chain.Config().IsShanghai(header.TimeBig())
+	if shanghai && header.WithdrawalsHash == nil {
+		return fmt.Errorf("missing withdrawalsHash")
+	}
+	// Verify existence / non-existence of withdrawalsHash.
+	if !shanghai && header.WithdrawalsHash != nil {
+		return fmt.Errorf("invalid withdrawalsHash: have %s, expected nil", header.WithdrawalsHash)
+	}
+	if chain.Config().IsSharding(header.TimeBig()) {
```

Comment:
I think we don't check that excessdataGas is not set before sharding
``` 
if !chain.Config().IsSharding(header.TimeBig()) {
		if header.ExcessDataGas != nil {
			return fmt.Errorf("invalid excessDataGas before fork: have %v, want <nil>", header.ExcessDataGas)
		}
	} else if err := misc.VerifyEip4844Header(chain.Config(), parent, header); err != nil {
		// Verify the header's EIP-4844 attributes.
		return err
	}
```

Reply:
good catch, fixed